repository/src/main/java/org/alfresco/repo/security/permissions/impl/acegi/ACLEntryAfterInvocationProvider.java

```diff
@@ -269,7 +269,18 @@ else if (NodeRef.class.isAssignableFrom(returnedObject.getClass()))
             }
             else if (Pair.class.isAssignableFrom(returnedObject.getClass()))
             {
-                return decide(authentication, object, config, (Pair) returnedObject);
+                Pair<?, ?> pair = (Pair<?, ?>) returnedObject;
+                if (pair.getSecond() != null && NodeRef.class.isAssignableFrom(pair.getSecond().getClass()))
+                {
+                    return decide(authentication, object, config, pair);
+                } else
+                {
+                    if (log.isDebugEnabled())
+                    {
+                        log.debug("Uncontrolled object - access allowed for " + object.getClass().getName());
+                    }
+                    return returnedObject;
+                }

```

Comment:
If we're adding extra logic around this then I think it should go into the `decide` method.

Reply:
Sounds good, done